src/main/java/com/google/devtools/build/lib/remote/RemoteExecutionService.java

```diff
@@ -507,8 +518,15 @@ public RemoteAction buildRemoteAction(Spawn spawn, SpawnExecutionContext context
       throws IOException, ExecException, ForbiddenActionInputException, InterruptedException {
     remoteActionBuildingSemaphore.acquire();
     try {
+      // Create a remote path resolver that is aware of the spawn's path mapper, which rewrites
+      // the paths of the inputs and outputs as well as paths appearing in the command line for
+      // execution. This is necessary to ensure that artifacts are correctly emitted into and staged
+      // from the unmapped location locally.
+      RemotePathResolver remotePathResolver =
+          RemotePathResolver.createMapped(baseRemotePathResolver, execRoot, spawn.getPathMapper());
       ToolSignature toolSignature = getToolSignature(spawn, context);
-      final MerkleTree merkleTree = buildInputMerkleTree(spawn, context, toolSignature);
+      final MerkleTree merkleTree =
+          buildInputMerkleTree(spawn, context, toolSignature, remotePathResolver);
```

Comment:
I'm confused. Doesn't `buildInputMerkleTree` already (indirectly) call `SpawnInputExpander.getInputMapping`, which (after #) takes care of applying the `PathMapper`?

Reply:
Oh, I see, this is only about the output tree artifact directories, which aren't inputs. Nevermind.